### ISSUE TITLE
Improve for_where rule

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#1387]
If the `if` is too complex, it's better to have it instead of a `where` clause.

Also, that bug that prevents us to recognize local variables sometimes makes the rule trigger false positives.

**Comment by keith** [comment]
I found a false positive for this rule as well:

```swift
for a in [1, 2, 3] {
    if a == 1 {
        return 0
    }

    return a
}
```

Here the `a == 1` could not be added as a `where` since we still want to do something in the alternate case. A simple workaround for now:

```swift
for a in [1, 2, 3] {
    if a == 1 {
        return 0
    } else {
        return a
    }
}
```

**Comment by marcelofabri** [comment]
@keith Could you try running with #1393? I think that should fix the issue as well

**Comment by keith** [comment]
Trying!

**Comment by keith** [comment]
Confirmed! Thanks you beat me to it!!

**Comment by eaigner** [comment]
Another false positive

![screen shot 2019-02-08 at 13 15 09](https://user-images.githubusercontent.com/112197/52477863-b8b45380-2ba3-11e9-8861-611053184645.png)

This rule should not apply if the if block `return`s.

**Comment by jpsim** [comment]
That's not a false positive, these are all equivalent:

```swift
private var allFilesExpanded: Bool {
    for index in visibleFileRowIndexes {
        if !visibleParts[index].isExpanded {
            return false
        }
    }
    return true
}
```

```swift
private var allFilesExpanded: Bool {
    for index in visibleFileRowIndexes where !visibleParts[index].isExpanded {
        return false
    }
    return true
}
```

and finally:

```swift
private var allFilesExpanded: Bool {
    return visibleFileRowIndexes.allSatisfy { visibleParts[$0].isExpanded }
}
```
